airflow/models/dagrun.py

```diff
@@ -810,20 +822,27 @@ def verify_integrity(self, session: Session = NEW_SESSION):
             task = None
             try:
                 task = dag.get_task(ti.task_id)
+
+                should_restore_task = (task is not None) and ti.state == State.REMOVED
+                if should_restore_task:
+                    self.log.info("Restoring task '%s' which was previously removed from DAG '%s'", ti, dag)
+                    Stats.incr(f"task_restored_to_dag.{dag.dag_id}", 1, 1)
+                    ti.state = State.NONE
             except AirflowException:
                 if ti.state == State.REMOVED:
                     pass  # ti has already been removed, just ignore it
                 elif self.state != State.RUNNING and not dag.partial:
                     self.log.warning("Failed to get task '%s' for dag '%s'. Marking it as removed.", ti, dag)
                     Stats.incr(f"task_removed_from_dag.{dag.dag_id}", 1, 1)
                     ti.state = State.REMOVED
+                continue
 
-            should_restore_task = (task is not None) and ti.state == State.REMOVED
-            if should_restore_task:
-                self.log.info("Restoring task '%s' which was previously removed from DAG '%s'", ti, dag)
-                Stats.incr(f"task_restored_to_dag.{dag.dag_id}", 1, 1)
-                ti.state = State.NONE
-            session.merge(ti)
```

Comment:
This `session.merge()` call is gone in the new implementation; is it intended?

Reply:
Yes -- the TIs are fetched from the current attached session so this shouldn't be required?